packages/gateway/src/utils/cid.js

```diff
@@ -0,0 +1,34 @@
+import { CID } from 'multiformats/cid'
+
+import { InvalidIpfsPathError, InvalidUrlError } from '../errors.js'
+
+/**
+ * Parse subdomain URL and return cid
+ *
+ * @param {URL} url
+ */
+export function getCidFromSubdomainUrl(url) {
+  // Replace "ipfs-staging" by "ipfs" if needed
+  const nUrl = url.hostname.replace('ipfs-staging', 'ipfs')
+  const splitUrl = nUrl.split('.ipfs.')
+
+  if (!splitUrl.length) {
+    throw new InvalidUrlError(url.hostname)
+  }
+
+  return normalizeCid(splitUrl[0])
```

Comment:
It's not really an invalid "IPFS path" it's an invalid CID in the hostname of the URL, I'd probably do:

```suggestion
  try {
    return normalizeCid(splitUrl[0])
  } catch (err) {
    throw new InvalidUrlError(`invalid CID: ${splitUrl[0]}: ${err.message}`)
  }
```

...you'd need to also change `InvalidUrlError` to accept a message instead of a URL.